libs/component-tests/src/main/java/org/codice/ddf/test/component/UrlBuilder.java

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.test.component;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UrlBuilder {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UrlBuilder.class);
+
+    private final URL rootContextUrl;
+
+    private final URL secureRootContextUrl;
+
+    public static UrlBuilder from(String hostname, String port, String securePort,
+            String rootContext) {
+        return new UrlBuilder(hostname, port, securePort, rootContext);
+    }
+
+    public static UrlBuilder fromSystemProperties(String hostnameSystemProperty,
+            String portSystemProperty, String securePortSystemProperty,
+            String rootContextSystemProperty) {
+
+        return new UrlBuilder(System.getProperty(hostnameSystemProperty),
+                System.getProperty(portSystemProperty),
+                System.getProperty(securePortSystemProperty),
+                System.getProperty(rootContextSystemProperty));
+    }
+
+    public UrlBuilder add(String contextPath) {
+        String cleanContextPath = cleanContextPath(contextPath);
+
+        try {
+            return new UrlBuilder(//
```

Comment:
❓ This seems like more manipulation than necessary to build up the URL. Why not just keep track of the context path and build the secure or public URL when it is requested? If you wanted to get really fancy you could even cache the URL and invalidate it when `add()` is called. For sanity you could prohibit slashes, e.g. `add(services).add(catalog)...` That feels more buildery to me anyways.

This is probably premature optimization, but it seems like it could be an easy thing to trip over in loops or as we start to use this class more throughout the codebase.

Reply:
Good input @brendan-hofmann. I wanted to put that out there so we can start iterating on it. I'm all for improving it but don't want to hold up this PR any longer.